pkg/agent/controller/networkpolicy/l7engine/reconciler.go

```diff
@@ -0,0 +1,398 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	v1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+)
+
+const (
+	defaultSuricataConfigPath = "/etc/suricata/suricata.yaml"
+	antreaSuricataConfigPath  = "/etc/suricata/antrea.yaml"
+
+	tenantConfigsDir = "/etc/suricata"
+	tenantRulesDir   = "/etc/suricata/rules"
+
+	suricataCommandSocket = "/var/run/suricata/suricata-command.socket"
+
+	addressAny = "any"
+	portAny    = "any"
+
+	actionPass   = "pass"
+	actionReject = "reject"
+
+	protocolHTTP = "http"
+	protocolIP   = "ip"
+)
+
+var (
+	// Declared as a variable for testing.
+	defaultFS = afero.NewOsFs()
+)
+
+type Reconciler struct {
+	// Declared as member variables for testing.
+	startSuricataFn func()
+	suricataScFn    func(scCmd string) (string, error)
+
+	once sync.Once
+}
+
+func NewReconciler() *Reconciler {
+	r := &Reconciler{
+		suricataScFn:    suricataSc,
+		startSuricataFn: startSuricata,
+	}
+
+	return r
+}
+
+func generateRulesConf(policyName, ipSrc, ipDst, tpDst string, protoKeywords map[string]sets.String) *bytes.Buffer {
+	rulesBuffer := bytes.NewBuffer(nil)
+	sid := 1
+
+	// Generate default reject rule.
+	allKeywords := fmt.Sprintf(`msg: "Reject dy default reject rule"; flow: to_server, established; sid: %d;`, sid)
+	rule := fmt.Sprintf("%s %s %s %s -> %s %s (%s)\n", actionReject, protocolIP, addressAny, portAny, addressAny, portAny, allKeywords)
+	rulesBuffer.WriteString(rule)
+	sid++
+
+	// Generate rules from rulesInfoMap.
+	for proto, keywordsSet := range protoKeywords {
+		for keywords := range keywordsSet {
+			if ipSrc == "" || ipDst == "" {
+				// Without source or destination IP addresses, don't generate a rule.
+				continue
+			}
+			// It is a convention that the sid is provided as the last keyword (or second-to-last if there is a rev)
+			// of a rule.
+			if keywords != "" {
+				allKeywords = fmt.Sprintf(`msg: "Allow %s by %s"; %s sid: %d;`, proto, policyName, keywords, sid)
+			} else {
+				allKeywords = fmt.Sprintf(`msg: "Allow %s by %s"; sid: %d;`, proto, policyName, sid)
+			}
+			rule = fmt.Sprintf("%s %s %s %s -> %s %s (%s)\n", actionPass, proto, ipSrc, portAny, ipDst, tpDst, allKeywords)
+			rulesBuffer.WriteString(rule)
+			sid++
+		}
+	}
+
+	return rulesBuffer
+}
+
+func generateRulesPath(vlanID uint32) string {
+	return fmt.Sprintf("%s/antrea-l7-networkpolicy-%d.rules", tenantRulesDir, vlanID)
+}
+
+func generateTenantConfigPath(vlanID uint32) string {
+	return fmt.Sprintf("%s/antrea-tenant-%d.yaml", tenantConfigsDir, vlanID)
+}
+
+func writeConfigFile(path string, data *bytes.Buffer) error {
+	f, err := defaultFS.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err = f.Write(data.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func convertAddresses(group v1beta.GroupMemberSet) string {
+	// Without any group, use 'any'.
+	if group == nil {
+		return addressAny
+	}
```

Comment:
1. This can never happen.
2. From.IPBlock and To.IPBlock should be counted in too. otherwise it wouldn't work when IPBlock is used.
3. Since only matched traffic would be redirected to this tenant now, is it still necessary to check ip and port? If no, I think it could be much simpler and it won't be limited by the maximum of IP addresses in a signature even with Suricata 6.0.

Reply:
You reminded me. I think we can remove source/destination IP addresses and ports in Suricata rule since every rule has its own tenant. It's unnecessary to filter L3 and L4 again in Suricata.